# Step 4 and 5 Pre-Processing and Modeling

   **Pre-processing and Training Data Development**  
 * Create dummy or indicator features for categorical variables
  * Standardize the magnitude of numeric features
  * Split into testing and training datasets
  * Apply scaler to the testing set

**<font color='teal'> Start by loading the necessary packages as we did in step 3 and printing out our current working directory just to confirm we are in the correct project directory. </font>**

In [2]:
#load python packages
import os
import pandas as pd
import datetime
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
%matplotlib inline

In [3]:
path = os.getcwd()

**<font color='teal'>  Load the csv file you created in step 3</font>**

In [418]:
directory = "data"
data_dir = os.path.join(path, directory)
df = pd.read_csv(os.path.join(data_dir,'step3_output.csv'),sep='\t')

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `state`. Add the dummies back to the dataframe and remove the original column for `state`. </font>**

In [419]:
dfo=df[['state']] # select object type columns
df = pd.concat([df.drop(dfo, axis=1), pd.get_dummies(dfo)], axis=1)

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing standardize the scale of the features of the dataframe except the name of the resort which we don't need in the dataframe for modeling, so it can be droppped here as well. Also, we want to hold out our response variable(s) so we can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as our response for scaling and modeling.</font>**

In [420]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df.drop(['Name','AdultWeekend','clusters'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df['AdultWeekend'] 

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 

## Split into training and testing datasets

**<font color='teal'> Using sklearn model selection import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. We will start by using the adult weekend ticket price as our response variable for modeling.</font>**

In [421]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()


# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

Here we start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Modeling


Now it's time to get into the most exciting part: modeling! In this exercise, we will build three different models and compare each model's performance. In the end, we will choose the best model for demonstrating insights to Big Mountain management.

  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

## Fit Models with a Training Dataset

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [422]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

In [423]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred = model.predict(X_test)

## Review Model Outcomes — Iterate over additional models as needed

In [424]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics. 
# You can plug y_test and y_pred into the functions to evaluate the model
from sklearn.metrics import explained_variance_score
from sklearn.metrics import mean_absolute_error

In [425]:
ev = explained_variance_score(y_test, y_pred)
mae = mean_absolute_error(y_test, y_pred)

In [426]:
print("explained_vaiance_score: ",ev)
print("mean_absolute_error: ",mae)

explained_vaiance_score:  -1.1691191971732987e+23
mean_absolute_error:  936755800739.2261


**<font color='teal'> Print the intercept value from the linear model. </font>**

Hint: our linear regression model `lm` has an attribute `intercept_` for the intercept

In [427]:
print(lm.intercept_)

236324411799.34882


**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the resulting `AdultWeekend` value. Also, because we took the time to scale our x values in the training data, we can compare each of the coeeficients for the features to determine the feature importances. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 


Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [428]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
df_lm = pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])

In [429]:
print(df_lm.sort_values('Coefficient', ascending=False))

                       Coefficient
fastEight             4.105281e+12
total_chairs          1.778877e+12
fastQuads             6.743129e+11
surface               6.293894e+11
double                5.560638e+11
state_New York        5.228692e+11
triple                4.961700e+11
state_Michigan        4.856825e+11
state_Colorado        4.347812e+11
state_California      4.254762e+11
state_Pennsylvania    4.060204e+11
quad                  4.023439e+11
state_Wisconsin       3.743884e+11
state_New Hampshire   3.743884e+11
state_Vermont         3.630786e+11
state_Minnesota       3.513254e+11
state_Utah            3.390826e+11
state_Montana         3.262950e+11
state_Idaho           3.262950e+11
state_Massachusetts   3.128961e+11
state_Oregon          2.988034e+11
state_Washington      2.988034e+11
state_New Mexico      2.839138e+11
state_Maine           2.839138e+11
state_Wyoming         2.680944e+11
state_North Carolina  2.328987e+11
state_Ohio            2.129353e+11
state_Connecticut   

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that we care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

Hint: Try to construct another model using exactly the steps we followed above. 

#### Model 2

In [435]:
# drop state column from dataframe
df2 = pd.read_csv(os.path.join(data_dir,'step3_output.csv'),sep='\t')
X2 = df2.drop(['state','Name','AdultWeekend'], axis=1)

In [436]:
# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scalar2 = preprocessing.StandardScaler().fit(X2)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X2_scaled=scalar2.transform(X2) 

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X2_train, X2_test, y2_train, y2_test = train_test_split(X2_scaled, y, test_size=0.25, random_state=1)

In [437]:
# modeling
lm2 = linear_model.LinearRegression()
model2 = lm2.fit(X2_train,y2_train)

In [438]:
# Make a variable called y2_pred and assign it the result of calling predict() on our model variable with parameter X_test
y2_pred = model2.predict(X2_test)

In [439]:
ev2 = explained_variance_score(y2_test, y2_pred)
mae2 = mean_absolute_error(y2_test, y2_pred)

print("explained_vaiance_score: ",ev2)
print("mean_absolute_error: ",mae2)

explained_vaiance_score:  0.897895869840222
mean_absolute_error:  5.804824974000846


In [440]:
print(lm2.intercept_)

64.12769707219066


In [441]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
df2_lm = pd.DataFrame(abs(lm2.coef_), X2.columns, columns=['Coefficient'])

In [442]:
print(df2_lm.sort_values('Coefficient', ascending=False))

                    Coefficient
AdultWeekday       1.995936e+01
clusters           6.311616e+00
summit_elev        4.611518e+00
SkiableTerrain_ac  2.181497e+00
vertical_drop      1.882141e+00
NightSkiing_ac     1.605362e+00
Snow Making_ac     1.535660e+00
surface            1.281830e+00
fastSixes          1.176145e+00
fastQuads          1.126889e+00
Runs               1.084779e+00
triple             9.877901e-01
quad               8.441318e-01
TerrainParks       7.756796e-01
trams              7.408327e-01
total_chairs       5.705196e-01
LongestRun_mi      4.045848e-01
double             2.232545e-01
daysOpenLastYear   1.238936e-01
yearsOpen          1.181421e-01
averageSnowfall    1.050757e-01
projectedDaysOpen  1.864247e-02
fastEight          3.552714e-15


**<font color='teal'> When reviewing our new model coefficients, we see `summit_elev` is now in the number three spot. This is also difficult to change from a management prespective and highly correlated with `base_elev` and `vertical_drop`.  This time, rebuild the model without the state features and without the `summit_elev` and compare the results. </font>**

#### Model 3

In [443]:
# drop state column from dataframe
df3 = pd.read_csv(os.path.join(data_dir,'step3_output.csv'),sep='\t')
X3 = df3.drop(['state','Name','AdultWeekend','summit_elev'], axis=1)

In [444]:
# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scalar3 = preprocessing.StandardScaler().fit(X3)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X3_scaled=scalar3.transform(X3) 

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X3_train, X3_test, y3_train, y3_test = train_test_split(X3_scaled, y, test_size=0.25, random_state=1)

In [445]:
# modeling
lm3 = linear_model.LinearRegression()
model3 = lm3.fit(X3_train,y3_train)

In [446]:
# Make a variable called y2_pred and assign it the result of calling predict() on our model variable with parameter X_test
y3_pred = model3.predict(X3_test)

In [447]:
ev3 = explained_variance_score(y3_test, y3_pred)
mae3 = mean_absolute_error(y3_test, y3_pred)

print("explained_vaiance_score: ",ev3)
print("mean_absolute_error: ",mae3)

explained_vaiance_score:  0.8953112166541771
mean_absolute_error:  5.7197803110671765


In [448]:
print(lm3.intercept_)

64.11101380678544


In [449]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
df3_lm = pd.DataFrame(abs(lm3.coef_), X3.columns, columns=['Coefficient'])

In [450]:
print(df3_lm.sort_values('Coefficient', ascending=False))

                    Coefficient
AdultWeekday       2.024011e+01
clusters           2.992526e+00
vertical_drop      2.784703e+00
Snow Making_ac     1.816696e+00
fastSixes          1.235781e+00
surface            1.174397e+00
Runs               1.145277e+00
NightSkiing_ac     1.115082e+00
SkiableTerrain_ac  1.096735e+00
quad               1.082895e+00
triple             1.041710e+00
TerrainParks       9.626981e-01
trams              9.620400e-01
fastQuads          8.378244e-01
daysOpenLastYear   6.433748e-01
total_chairs       5.675776e-01
projectedDaysOpen  5.275613e-01
LongestRun_mi      3.887476e-01
double             2.688058e-01
averageSnowfall    2.582642e-01
yearsOpen          6.264172e-02
fastEight          1.554312e-15


## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell — you will discuss this selection more in the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. | -1.1691191971732987e+23 |936755800739.2261  ||
| Model 2. | 0.897895869840222|5.804824974000846 |'state'|
| Model 3. | 0.8953112166541771 |5.7197803110671765 |'state','summit_elev'|

Model Selection:

In [0]:
model3